// carbon-components-react/lib/components/Grid/Grid-story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fullWidth = exports.condensedColumns = exports.condensed = exports.offset = exports.responsiveGrid = exports.autoColumns = exports.default = void 0;

require("./Grid-story.scss");

var _react = _interopRequireDefault(require("react"));

var _ = require("./");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  decorators: [function (storyFn) {
    return /*#__PURE__*/_react.default.createElement("div", {
      id: "templates"
    }, storyFn());
  }],
  title: 'Grid'
};
exports.default = _default;

function DemoContent(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "outside"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "inside"
  }, children));
}

var autoColumns = function autoColumns() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, null, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))));
};

exports.autoColumns = autoColumns;

var responsiveGrid = function responsiveGrid() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, null, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: 1,
    md: 4,
    lg: 8
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "sm: 1/4, md: 1/2, lg: 2/3")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: 1,
    md: 2,
    lg: 2
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "sm: 1/4, md: 1/4, lg: 1/6")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: 1,
    md: 1,
    lg: 1
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "sm: 1/4, md: 1/8, lg: 1/12")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: 1,
    md: 1,
    lg: 1
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "sm: 1/4, md: 1/8, lg: 1/12"))));
};

exports.responsiveGrid = responsiveGrid;

var offset = function offset() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, null, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: {
      span: 1,
      offset: 3
    }
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "Small Screen Offset 3")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: {
      span: 2,
      offset: 2
    }
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "Small Screen Offset 2")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: {
      span: 3,
      offset: 1
    }
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "Small Screen Offset 1")), /*#__PURE__*/_react.default.createElement(_.Column, {
    sm: {
      span: 4,
      offset: 0
    }
  }, /*#__PURE__*/_react.default.createElement(DemoContent, null, "Small Screen Offset 0"))));
};

exports.offset = offset;

var condensed = function condensed() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, {
    condensed: true
  }, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))));
};

exports.condensed = condensed;

var condensedColumns = function condensedColumns() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, null, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))), /*#__PURE__*/_react.default.createElement(_.Row, {
    condensed: true
  }, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))), /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))));
};

exports.condensedColumns = condensedColumns;

var fullWidth = function fullWidth() {
  return /*#__PURE__*/_react.default.createElement(_.Grid, {
    fullWidth: true
  }, /*#__PURE__*/_react.default.createElement(_.Row, null, /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4")), /*#__PURE__*/_react.default.createElement(_.Column, null, /*#__PURE__*/_react.default.createElement(DemoContent, null, "1/4"))));
};

exports.fullWidth = fullWidth;
