// carbon-components-react/es/components/Grid/Grid-story.js
import './Grid-story.scss';
import React from 'react';
import { Grid, Row, Column } from './';
export default {
  decorators: [function (storyFn) {
    return /*#__PURE__*/React.createElement("div", {
      id: "templates"
    }, storyFn());
  }],
  title: 'Grid'
};

function DemoContent(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement("div", {
    className: "outside"
  }, /*#__PURE__*/React.createElement("div", {
    className: "inside"
  }, children));
}

export var autoColumns = function autoColumns() {
  return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))));
};
export var responsiveGrid = function responsiveGrid() {
  return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, {
    sm: 1,
    md: 4,
    lg: 8
  }, /*#__PURE__*/React.createElement(DemoContent, null, "sm: 1/4, md: 1/2, lg: 2/3")), /*#__PURE__*/React.createElement(Column, {
    sm: 1,
    md: 2,
    lg: 2
  }, /*#__PURE__*/React.createElement(DemoContent, null, "sm: 1/4, md: 1/4, lg: 1/6")), /*#__PURE__*/React.createElement(Column, {
    sm: 1,
    md: 1,
    lg: 1
  }, /*#__PURE__*/React.createElement(DemoContent, null, "sm: 1/4, md: 1/8, lg: 1/12")), /*#__PURE__*/React.createElement(Column, {
    sm: 1,
    md: 1,
    lg: 1
  }, /*#__PURE__*/React.createElement(DemoContent, null, "sm: 1/4, md: 1/8, lg: 1/12"))));
};
export var offset = function offset() {
  return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, {
    sm: {
      span: 1,
      offset: 3
    }
  }, /*#__PURE__*/React.createElement(DemoContent, null, "Small Screen Offset 3")), /*#__PURE__*/React.createElement(Column, {
    sm: {
      span: 2,
      offset: 2
    }
  }, /*#__PURE__*/React.createElement(DemoContent, null, "Small Screen Offset 2")), /*#__PURE__*/React.createElement(Column, {
    sm: {
      span: 3,
      offset: 1
    }
  }, /*#__PURE__*/React.createElement(DemoContent, null, "Small Screen Offset 1")), /*#__PURE__*/React.createElement(Column, {
    sm: {
      span: 4,
      offset: 0
    }
  }, /*#__PURE__*/React.createElement(DemoContent, null, "Small Screen Offset 0"))));
};
export var condensed = function condensed() {
  return /*#__PURE__*/React.createElement(Grid, {
    condensed: true
  }, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))));
};
export var condensedColumns = function condensedColumns() {
  return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))), /*#__PURE__*/React.createElement(Row, {
    condensed: true
  }, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))));
};
export var fullWidth = function fullWidth() {
  return /*#__PURE__*/React.createElement(Grid, {
    fullWidth: true
  }, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4")), /*#__PURE__*/React.createElement(Column, null, /*#__PURE__*/React.createElement(DemoContent, null, "1/4"))));
};
