// carbon-components-react/es/components/DataTable/stories/with-overflow-menu.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import OverflowMenu from '../../OverflowMenu';
import OverflowMenuItem from '../../OverflowMenuItem';
import DataTable, { Table, TableBody, TableCell, TableContainer, TableHead, TableHeader, TableRow, TableSelectAll, TableSelectRow } from '../../DataTable';
import { initialRows, headers } from './shared';

var OverflowMenuStory = function OverflowMenuStory(props) {
  return /*#__PURE__*/React.createElement(DataTable, _extends({
    rows: initialRows,
    headers: headers
  }, props, {
    render: function render(_ref) {
      var rows = _ref.rows,
          headers = _ref.headers,
          getHeaderProps = _ref.getHeaderProps,
          getRowProps = _ref.getRowProps,
          getSelectionProps = _ref.getSelectionProps,
          getTableProps = _ref.getTableProps;
      return /*#__PURE__*/React.createElement(TableContainer, {
        title: "DataTable",
        description: "With selection"
      }, /*#__PURE__*/React.createElement(Table, getTableProps(), /*#__PURE__*/React.createElement(TableHead, null, /*#__PURE__*/React.createElement(TableRow, null, /*#__PURE__*/React.createElement(TableSelectAll, getSelectionProps()), headers.map(function (header, i) {
        return /*#__PURE__*/React.createElement(TableHeader, _extends({
          key: i
        }, getHeaderProps({
          header: header
        })), header.header);
      }), /*#__PURE__*/React.createElement(TableHeader, null))), /*#__PURE__*/React.createElement(TableBody, null, rows.map(function (row, i) {
        return /*#__PURE__*/React.createElement(TableRow, _extends({
          key: i
        }, getRowProps({
          row: row
        })), /*#__PURE__*/React.createElement(TableSelectRow, getSelectionProps({
          row: row
        })), row.cells.map(function (cell) {
          return /*#__PURE__*/React.createElement(TableCell, {
            key: cell.id
          }, cell.value);
        }), /*#__PURE__*/React.createElement(TableCell, {
          className: "bx--table-column-menu"
        }, /*#__PURE__*/React.createElement(OverflowMenu, {
          flipped: true
        }, /*#__PURE__*/React.createElement(OverflowMenuItem, null, "Action 1"), /*#__PURE__*/React.createElement(OverflowMenuItem, null, "Action 2"), /*#__PURE__*/React.createElement(OverflowMenuItem, null, "Action 3"))));
      }))));
    }
  }));
};

export default OverflowMenuStory;
