// carbon-components-react/es/components/DataTable/stories/with-batch-expansion.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import DataTable, { Table, TableBody, TableCell, TableContainer, TableExpandHeader, TableExpandRow, TableExpandedRow, TableHead, TableHeader, TableRow } from '../../DataTable';
import { initialRows, headers } from './shared';
import './with-expansion-story.scss';

var BatchExpansionStory = function BatchExpansionStory(props) {
  return /*#__PURE__*/React.createElement(DataTable, _extends({
    rows: initialRows,
    headers: headers
  }, props, {
    render: function render(_ref) {
      var rows = _ref.rows,
          headers = _ref.headers,
          getHeaderProps = _ref.getHeaderProps,
          getExpandHeaderProps = _ref.getExpandHeaderProps,
          getRowProps = _ref.getRowProps,
          getTableProps = _ref.getTableProps,
          getTableContainerProps = _ref.getTableContainerProps;
      return /*#__PURE__*/React.createElement(TableContainer, _extends({
        title: "DataTable",
        description: "With expansion"
      }, getTableContainerProps()), /*#__PURE__*/React.createElement(Table, getTableProps(), /*#__PURE__*/React.createElement(TableHead, null, /*#__PURE__*/React.createElement(TableRow, null, /*#__PURE__*/React.createElement(TableExpandHeader, _extends({
        enableExpando: true
      }, getExpandHeaderProps())), headers.map(function (header, i) {
        return /*#__PURE__*/React.createElement(TableHeader, _extends({
          key: i
        }, getHeaderProps({
          header: header
        })), header.header);
      }))), /*#__PURE__*/React.createElement(TableBody, null, rows.map(function (row) {
        return /*#__PURE__*/React.createElement(React.Fragment, {
          key: row.id
        }, /*#__PURE__*/React.createElement(TableExpandRow, getRowProps({
          row: row
        }), row.cells.map(function (cell) {
          return /*#__PURE__*/React.createElement(TableCell, {
            key: cell.id
          }, cell.value);
        })), /*#__PURE__*/React.createElement(TableExpandedRow, {
          colSpan: headers.length + 1,
          className: "demo-expanded-td"
        }, /*#__PURE__*/React.createElement("h6", null, "Expandable row content"), /*#__PURE__*/React.createElement("div", null, "Description here")));
      }))));
    }
  }));
};

export default BatchExpansionStory;
